include/openthread/platform/dso_transport.h

```diff
@@ -0,0 +1,202 @@
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *   This file includes the platform abstraction for DNS Stateful Operations (DSO) transport.
+ */
+
+#ifndef OPENTHREAD_PLATFORM_DSO_TRANSPORT_H_
+#define OPENTHREAD_PLATFORM_DSO_TRANSPORT_H_
+
+#include <stdint.h>
+
+#include <openthread/error.h>
+#include <openthread/instance.h>
+#include <openthread/ip6.h>
+#include <openthread/message.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * This structure represents a DSO connection.
+ *
+ * It is an opaque struct (the platform implementation only deals with pointers to this struct).
+ *
+ */
+typedef struct otPlatDsoConnection otPlatDsoConnection;
+
+/**
+ * This function can be used by DSO platform implementation to get the the OpenThread instance associated with a
+ * connection instance.
+ *
+ * @param[in] aConnection   A pointer to the DSO connection.
+ *
+ * @returns A pointer to the `otInstance`.
+ *
+ */
+extern otInstance *otPlatDsoGetInstance(otPlatDsoConnection *aConnection);
+
+/**
+ * This function starts or stops listening for incoming connection requests on transport layer.
+ *
+ * For DNS-over-TLS, the transport layer MUST listen on port 853 and follow RFC 7858.
+ *
+ * While listening is enabled, if a connection request is received, the `otPlatDsoAccept()` callback MUST be called.
+ *
+ * @param[in] aInstance    The OpenThread instance
+ * @param[in] aEnable      TRUE to start listening, FALSE to stop listening.
+ *
+ */
+void otPlatDsoEnableListening(otInstance *aInstance, bool aEnabled);
```

Comment:
DSO can operate on different transports, over TCP or over TLS.

SRPL seems to require DNS-over-TLS which is on port 853.  

The DSO implementation is currently aimed for SRPL as its first use-case but I guess it can be extended to handle other use-cases when/if needed.

> Another potential issue I can see is that platform will need to use the same listening socket for both DSO and non-DSO DNS over TCP connections.

That's correct, but I guess it is possible to do so. This API is just requesting to start listening for connection requests from others on DSO transport (on a single socket or may be even on multiple sockets). Platform may have already a socket open for other operations (say for DNS-over-TCP) and then reuse the existing open socket for DSO as well.